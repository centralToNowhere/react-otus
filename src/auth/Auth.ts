import { IPlayer } from "@/player/Player";

export interface IAuthState {
  player: IPlayer;
  loginPending: boolean;
  loginError: string;
}

export type LoginError = {
  message: string;
};

export const PlayerNameNotSetError: LoginError = new Error(
  "Player name is not specified!"
);

export const PlayerNotRegisteredError: LoginError = new Error(
  "Player is not registered somehow!"
);

export const registerPlayer = (playerName: string) => {
  return new Promise<IPlayer>((resolve, reject) => {
    setTimeout(() => {
      if (playerName) {
        resolve({
          registered: true,
          name: playerName,
        });
      } else {
        reject(PlayerNameNotSetError);
      }
    }, fakeRegisterPlayerTimeout);
  });
};

export const fakeRegisterPlayerTimeout = 500;
