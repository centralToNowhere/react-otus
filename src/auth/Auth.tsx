import React, {FC, useEffect, useState} from "react";
import { Navigate } from "react-router-dom";
import {AppAction, IAppState, IPlayer} from "@/state";
import {SetPlayerAction} from "@/state/actions";

export const storageKeyPlayerName = "playerName";
export const authProtected = <P extends { player: IPlayer }>(Component: React.ComponentType<P>): FC<P> => {
  const AuthProtected: FC<P> = (props) => {
    const { player } = props;

    useEffect(() => {
    }, [player.registered])
    return player.registered ? <Component {...props} /> : <Navigate to="/registration" />;
  }

  return AuthProtected;
}

export const getPlayer = (): IPlayer | null => {
  const playerData = localStorage.getItem(storageKeyPlayerName);

  return playerData ? JSON.parse(playerData) : null;
}

export const setPlayer = (player: IPlayer) => {
  localStorage.setItem(storageKeyPlayerName, JSON.stringify(player));
}

export const unsetPlayer = () => {
  localStorage.removeItem(storageKeyPlayerName);
}

export const usePlayerRegistration = (
  player: IPlayer,
  dispatch: React.Dispatch<AppAction>
): (playerName: string | null) => void => {
  useEffect(() => {
    setPlayer(player);
  }, [player]);

  return (playerName: string | null) => {
    if (playerName) {
      dispatch(SetPlayerAction({
        registered: true,
        name: playerName
      }));
    }
  }
}