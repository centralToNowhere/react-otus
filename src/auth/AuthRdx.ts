import {createAsyncThunk, createSelector, createSlice} from "@reduxjs/toolkit";
import {defaultPlayer, IPlayer} from "@/player/Player";
import {RootState} from "@/store/redux/store";
import {getPlayerDataFromStorage} from "@/storage/Storage";

export interface IAuthState {
  player: IPlayer,
  loginPending: boolean
}

const playerDataFromStorage = getPlayerDataFromStorage();

const initialState: IAuthState = {
  player: playerDataFromStorage?.player || defaultPlayer,
  loginPending: false
};

export const login = createAsyncThunk<IPlayer, string>(
  "auth/login",
  (playerName, thunkAPI) => {
    return new Promise<IPlayer>((resolve, reject) => {
      setTimeout(() => {
        if (playerName) {
          resolve({
            registered: true,
            name: playerName
          });
        }
      }, 500);
    });
  }
);

export const authSlice = createSlice({
  name: "auth",
  initialState: {
    ...initialState
  },
  reducers: {
    logout: (state) => {
      state.player.registered = false;
      state.player.name = "";
    }
  },
  extraReducers: (builder) => {
    builder.addCase(login.pending, (state, action) => {
      state.player.registered = false;
      state.loginPending = true;
      state.player.name = "";
    });
    builder.addCase(login.fulfilled, (state, action) => {
      state.player.registered = action.payload.registered;
      state.loginPending = false;
      state.player.name = action.payload.name;
    });
    builder.addCase(login.rejected, (state, action) => {
      state.player.registered = false;
      state.loginPending = false;
      state.player.name = "";
    });
  }
});

const selectAuth = (state: RootState) => {
  return state.auth;
}

export const selectPlayer = createSelector(
  selectAuth,
  (auth: IAuthState) => auth.player
);

export const selectLoginPending = createSelector(
  selectAuth,
  (auth: IAuthState) => auth.loginPending
);

export const { logout } = authSlice.actions;