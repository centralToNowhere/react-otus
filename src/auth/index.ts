export {
  authSlice,
  login,
  logout,
  selectPlayer,
  selectLoginPending
} from "./AuthRdx";
export type { IAuthState } from "./AuthRdx";
export {
  withAuthProtection,
  usePlayerRegistration
} from "./Auth";
export {storageKey} from "@/storage/Storage";