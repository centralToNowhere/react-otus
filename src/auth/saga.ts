import { call, put, select, takeEvery } from "redux-saga/effects";
import { clearPlayerData, persistPlayer } from "@/storage";
import { IPlayer } from "@/player/Player";
import {
  login,
  loginFailed,
  loginSucceed,
  logout,
  PlayerNotRegisteredError,
  registerPlayer,
  selectPlayerName,
} from "@/auth";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const onLogin = function* (): Generator<any, void, any> {
  const playerName: string = yield select(selectPlayerName);
  let player: IPlayer;

  try {
    player = yield call(registerPlayer, playerName);
  } catch (e) {
    yield put(loginFailed(e as Error));
    return;
  }

  if (player.registered) {
    yield put(loginSucceed(player));
  } else {
    yield put(loginFailed(PlayerNotRegisteredError));
  }
};

export const authSaga = function* () {
  yield takeEvery(loginSucceed.type, persistPlayer);
  yield takeEvery([logout.type, loginFailed.type], clearPlayerData);
  yield takeEvery(login.type, onLogin);
};
