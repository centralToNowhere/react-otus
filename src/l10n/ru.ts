export const l10n = {
  maxWidthLabel: "Макс. ширина:",
  maxHeightLabel: "Макс. высота:",
  cellSizeLabel: "Размер ячейки:",
  capacityLabel: "Процент заполненности:",
  speedLabel: "Обновлений в секунду:",
  buttonStart: "Старт",
  buttonStop: "Стоп",
  buttonReset: "Сброс",

  // Экран решистрации
  gameHeading: "Conway's Game of Life",
  buttonStartGameAsPlayer: "Начать игру",

  // Контейнер игрока
  registerPlayerLabel: "Введите имя игрока",
  logoutButton: "Выйти",
  statistics: "Статистика",
  score: "Очки:",
  clicksCount: "Количество кликов:",
};
