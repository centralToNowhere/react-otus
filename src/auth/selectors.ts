import { RootState } from "@/store/redux/store";
import { createSelector } from "@reduxjs/toolkit";
import { IPlayer } from "@/player/Player";
import { IAuthState } from "@/auth";

const selectAuth = (state: RootState) => {
  return state.auth;
};
export const selectPlayer = createSelector(
  selectAuth,
  (auth: IAuthState) => auth.player
);
export const selectPlayerName = createSelector(
  selectPlayer,
  (player: IPlayer) => player.name
);
export const selectLoginPending = createSelector(
  selectAuth,
  (auth: IAuthState) => auth.loginPending
);
export const selectLoginError = createSelector(
  selectAuth,
  (auth: IAuthState) => auth.loginError
);
