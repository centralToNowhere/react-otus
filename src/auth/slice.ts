import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { defaultPlayer, IPlayer } from "@/player/Player";
import { getDataFromStorage } from "@/storage";
import { IAuthState, LoginError } from "@/auth";

const playerDataFromStorage = getDataFromStorage();

const initialState: IAuthState = {
  player: playerDataFromStorage?.player || defaultPlayer,
  loginPending: false,
  loginError: "",
};

export const authSlice = createSlice({
  name: "auth",
  initialState: {
    ...initialState,
  },
  reducers: {
    logout: (state) => {
      state.player.registered = false;
      state.loginPending = false;
      state.player.name = "";
      state.loginError = "";
    },
    login: (state, action: PayloadAction<string>) => {
      state.player.registered = false;
      state.loginPending = true;
      state.player.name = action.payload;
      state.loginError = "";
    },
    loginSucceed: (state, action: PayloadAction<IPlayer>) => {
      state.player.registered = action.payload.registered;
      state.loginPending = false;
      state.player.name = action.payload.name;
      state.loginError = "";
    },
    loginFailed: (state, action: PayloadAction<LoginError>) => {
      state.player.registered = false;
      state.loginPending = false;
      state.player.name = "";
      state.loginError = action.payload.message;
    },
  },
});

export const { logout, login, loginFailed, loginSucceed } = authSlice.actions;
