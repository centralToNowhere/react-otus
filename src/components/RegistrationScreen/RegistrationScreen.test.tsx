import React, { FC, useReducer } from "react";
import { render, screen, waitFor } from "@testing-library/react";
import {
  IRegistrationScreenProps,
  RegistrationScreen,
} from "./RegistrationScreen";
import { l10n } from "@/l10n/ru";
import {
  AppReducer,
  defaultPlayer,
  IAppState,
  initialState,
} from "@/state/AppReducer";

import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event/dist";
import { usePlayerRegistration, storageKeyPlayerName } from "@/auth/Auth";

const RegistrationScreenWithDispatch: FC<IRegistrationScreenProps> = (
  props
) => {
  const [state, dispatch] = useReducer(AppReducer, {
    ...initialState,
    player: props.player,
  } as IAppState);

  const onPlayerRegistration = props.onPlayerRegistration
    ? props.onPlayerRegistration
    : //eslint-disable-next-line react-hooks/rules-of-hooks
      usePlayerRegistration(state.player, dispatch);

  return (
    <BrowserRouter>
      <RegistrationScreen
        player={state.player}
        onPlayerRegistration={onPlayerRegistration}
      />
    </BrowserRouter>
  );
};

describe("Registration screen tests", () => {
  it("should render registration screen", () => {
    const { asFragment } = render(
      <RegistrationScreenWithDispatch player={defaultPlayer} />
    );

    const header: HTMLHeadingElement = screen.getByText(l10n.gameHeading);
    const playerNameInput: HTMLInputElement = screen.getByLabelText(
      l10n.registerPlayerLabel
    );
    const gameStartButton: HTMLButtonElement = screen.getByText(
      l10n.buttonStartGameAsPlayer
    );

    expect(playerNameInput).toBeInTheDocument();
    expect(gameStartButton).toBeInTheDocument();
    expect(header).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("if player already registered, should show his name as input's initial value", async () => {
    render(
      <RegistrationScreenWithDispatch
        player={{
          registered: true,
          name: "Konstantin",
        }}
      />
    );

    const playerNameInput: HTMLInputElement = screen.getByLabelText(
      l10n.registerPlayerLabel
    );

    await waitFor(() => {
      expect(playerNameInput.value).toBe("Konstantin");
    });
  });

  it("should call onPlayerRegistration callback on button click", async () => {
    const callback = jest.fn();
    render(
      <RegistrationScreenWithDispatch
        onPlayerRegistration={callback}
        player={defaultPlayer}
      />
    );

    const gameStartButton: HTMLButtonElement = screen.getByText(
      l10n.buttonStartGameAsPlayer
    );

    userEvent.click(gameStartButton);

    await waitFor(() => {
      expect(callback).toHaveBeenCalledTimes(1);
    });
  });

  it("should call onPlayerRegistration callback on button enter", async () => {
    const callback = jest.fn();
    render(
      <RegistrationScreenWithDispatch
        onPlayerRegistration={callback}
        player={defaultPlayer}
      />
    );

    const gameStartButton: HTMLButtonElement = screen.getByText(
      l10n.buttonStartGameAsPlayer
    );

    userEvent.type(gameStartButton, "{enter}");

    await waitFor(() => {
      expect(callback).toHaveBeenCalledTimes(2);
    });
  });

  it("should save player data to localStorage after player registration", async () => {
    const storageSetItemSpy = jest.spyOn(Storage.prototype, "setItem");

    const player = {
      registered: true,
      name: "Konstantin",
    };

    render(<RegistrationScreenWithDispatch player={defaultPlayer} />);

    const playerNameInput: HTMLInputElement = screen.getByLabelText(
      l10n.registerPlayerLabel
    );
    const gameStartButton: HTMLButtonElement = screen.getByText(
      l10n.buttonStartGameAsPlayer
    );

    userEvent.type(playerNameInput, "Konstantin");
    userEvent.click(gameStartButton);

    await waitFor(() => {
      expect(storageSetItemSpy).toHaveBeenCalledWith(
        storageKeyPlayerName,
        JSON.stringify(player)
      );
    });
  });

  it("should not render game field if player name not set", async () => {
    render(<RegistrationScreenWithDispatch player={defaultPlayer} />);

    const gameStartButton: HTMLButtonElement = screen.getByText(
      l10n.buttonStartGameAsPlayer
    );

    userEvent.click(gameStartButton);

    const field = screen.queryByTestId("field");

    await waitFor(() => {
      expect(field).toBe(null);
    });
  });
});
