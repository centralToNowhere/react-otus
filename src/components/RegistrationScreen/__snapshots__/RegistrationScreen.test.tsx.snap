// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration screen tests should render registration screen 1`] = `
<DocumentFragment>
  <div
    class="css-14zx4vh"
  >
    <div
      class="css-1q1o94m"
    />
    <div
      class="css-2xdsfc"
    >
      <h1>
        Conway's Game of Life
      </h1>
      <form
        class="css-11przyw-PlayerRegistrationForm"
      >
        <div
          class="css-17wa1du-PlayerRegistrationForm"
        >
          <div
            class="css-1avd8c4-PlayerRegistrationForm"
          >
            <label
              class="css-1wnvjz2"
              for="register-player-input"
            >
              Введите имя игрока
            </label>
            <input
              class="css-1epispg-PlayerRegistrationForm"
              id="register-player-input"
              name="playerName"
              value=""
            />
          </div>
          <button
            class="css-14tin8v-PlayerRegistrationForm"
            name="startGameAsPlayer"
            type="button"
          >
            Начать игру
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
